### PR TITLE
[browser] Add tabs for mode private. Contributes to JB#50860

### PR DESCRIPTION
**Pull request by xlise**

### CODE REVIEW

File: apps/browser/qml/pages/components/TabView.qml
Comment (llewelld) [inline apps/browser/qml/pages/components/TabView.qml:61]: Currently the private mode background texture appears/disappears across the whole page as the tab is changed. I'm wondering if maybe it would look better if it slid in along with the page?

File: apps/browser/qml/pages/components/TabView.qml
Comment (llewelld) [inline apps/browser/qml/pages/components/TabView.qml:77]: Using the same `tabModel` and switching between them like this seems to be causing problems for the animation when swiping between the normal and private tab listings. The items appear suddenly half way through the swipe.

To get this to work nicely, I think it'll be necessary to maintain both models simultaneously and separately here.

File: apps/browser/qml/pages/components/TabView.qml
Comment (llewelld) [inline apps/browser/qml/pages/components/TabView.qml:21]: There's an error from Overlay.qml attempting to access `Browser.TabView.positionViewAtIndex()` It doesn't look like it's being caused by your changes here, but do you think it would be possible to fix it along with these changes?
```
[W] unknown:656 - file:///usr/share/sailfish-browser/pages/components/Overlay.qml:656: ReferenceError: positionViewAtIndex is not defined
[W] unknown:665 - file:///usr/share/sailfish-browser/pages/components/Overlay.qml:665: TypeError: Property 'positionViewAtIndex' of object TabView_QMLTYPE_144
```

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabGridView.qml:96]: Return to the previous page blocked, the menu is still left

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabGridView.qml:49]: It turned out to be problematic to fix, besides, this problem will go away when switching to the popup menu

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: done

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: I agree, corrected

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: I agree, corrected

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: Applied the hack, as there are no ideas how to do it correctly 
```
        Component.onCompleted:  {
            if ((index / columns * cellHeight) > tabView.height) {
                tabView.loaded = true
            }
        }
```

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (adenexter) [reply apps/browser/qml/pages/components/TabGridView.qml:96]: The gesture conflict could be reduced by disabling wrapping in the tab view. Then if you're on the first tab the gesture would return you to the previous page.

`wrapMode: PagedView.NoWrap` would do I think.

Comment (xlise) [reply apps/browser/qml/pages/components/TabGridView.qml:96]: Good idea, I will consult with the designer

**Comment by xlise** [comment]
Снимок_Экрана_20210520_001

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabGridView.qml:96]: Fixed
`interactive: false` It turned out like this

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (adenexter) [reply apps/browser/qml/pages/components/TabGridView.qml:49]: The convention is to put the menu above the header. The tab header background rectangle is going to create a bit of a problem there because it would need to be below the pull down menu so it doesn't occlude the highlight.

But for reference the way to position the menu at the top of the window would be to assign the tabs grid view to the TabItem flickable property and add a header item to the grid view that is the same height as TabView header.

```
Private.TabItem {
    flickable: _persistentTabView
    TabGridView {
         id: _persistentTabView
         header: Item {
              width: 1
              height: tabs.headerHeight
         }
   }
}
```

The horizontal offset might still be a problem even then though. Provided there's no clipping though, you might be able to fix it by adding a negative x to the PullDownMenu as a counter margin and increasing the width of the menu to resolve the margin on the right.

Alternatively including the margin in the GridView delegates instead of the view itself would work too.

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: Yes you are right. done

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabGridView.qml:50]: Thanks for your suggestions for improvement. Fixed gap. Moved the pulley menu above the title, but this can only be considered as a temporary solution before the implementation of the popup menu

**Comment by adenexter** [comment]
It's looking pretty good to me now. I am seeing a single pixel uneven line just below the pulley menu indicator. I think this is the normal tab header background item not being quite hidden by the black rectangle background and on just that one line you can see through to the web page below the page.

Hiding the tab header background with `_headerBackgroundVisible: false` in the silica TabView fixed that for me.

The scroll bar also gets hidden behind the tab header when the view is scrolled to the top but I don't have a solution for that, it is within a tab while needing to appear above a sibling item to the tab in the header.

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (llewelld) [inline apps/browser/qml/pages/components/TabGridView.qml:49]: Shifting the `SilicaGridView` down by `Theme.paddingLarge` here is unfortunately causing the pulley menu to be shifted down as well, so there's a gap at the top when you pull the items down.

I recommend removing this `y` shift here, and instead adding `Theme.paddingLarge` to both `TabGridView.header` items in `TabView.qml`.
```suggestion
```

File: apps/browser/qml/pages/components/TabView.qml
Comment (llewelld) [inline apps/browser/qml/pages/components/TabView.qml:131]: Similarly, the same extra space would need to be added here:
```suggestion
                        height: tabs.tabBarHeight + Theme.paddingLarge
```

File: apps/browser/qml/pages/components/TabView.qml
Comment (llewelld) [inline apps/browser/qml/pages/components/TabView.qml:98]: In order to compensate for removing the y offset from `TabGridView.qml` I suggest adding the extra space here instead.
```suggestion
                        height: tabs.tabBarHeight + Theme.paddingLarge
```

**Review by llewelld — CHANGES_REQUESTED** [review]
This is so much nicer, thank you. You even managed to solve the "swipe back" problem, which is fantastic. I've made just a few small comments, mostly about fixing some spacing issues. There also seem to be a couple of comments from Andrew still unresolved if I'm reading correctly.

**Comment by xlise** [comment]
> This is so much nicer, thank you. You even managed to solve the "swipe back" problem, which is fantastic. I've made just a few small comments, mostly about fixing some spacing issues. There also seem to be a couple of comments from Andrew still unresolved if I'm reading correctly.

Notes from Andrew will still edit

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: Thank you, done

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabGridView.qml]: Thank you, it's really neat. done

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabGridView.qml]: done

File: apps/browser/qml/pages/components/TabGridView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabGridView.qml]: done

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: done

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: done

**Comment by llewelld** [comment]
> The scroll bar also gets hidden behind the tab header when the view is scrolled to the top but I don't have a solution for that, it is within a tab while needing to appear above a sibling item to the tab in the header.

Maybe this is really nasty, but placing the `VerticalScrollBarDecorator` inside an `Item` that's only as high as the area beneath the toolbar could work. Something like this (this would need tweaking though because `tabsToolBar.height` isn't currently available inside the `TabViewGrid`s).

```
Item {
    y: tabsToolBar.height
    height: parent.height - y
    anchors.right: parent.right
    width: Math.round(Theme.paddingSmall/2)

    VerticalScrollDecorator {
        _forcedParent: parent
        flickable: tabView
    }
}
```

File: apps/browser/qml/pages/components/TabView.qml
Comment (llewelld) [inline apps/browser/qml/pages/components/TabView.qml:61]: I really like the way you've handled hiding the top menu highlight; it's very ingenious. However it seems to cause some issues, because `_pulleyMenuApparentHeight` doesn't update when the normal/private tab changes. I tried calling `_changeMenuApparentHeight()` when the `currentIndex` changed, but that resulted in other animation artefacts.

Perhaps as an alternative consider adding in a covering bar to the two `Private.TabItem` items instead. This way the bar will slide left/right when the grids slide. Something like this (just a rough idea):
```
Rectangle {
    color: Theme.colorScheme == Theme.LightOnDark ? "black" : "white"
    width: parent.width
    height: Math.max(0, _pulleyMenuHeight - y)
    y: Math.max(0, _pulleyMenuHeight - _persistentTabView.contentY + _persistentTabView.originY)
}   
```
It makes the code a bit messy, but then this is only a temporary solution until the menu is changed anyway.

Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml:61]: Thank you, so much better. done

Comment (llewelld) [reply apps/browser/qml/pages/components/TabView.qml:61]: Thanks @xlise. Unfortunately I don't see the change in practice (did you push it to the repo? I see a force-push, but I don't see any changes; it looks like a rebase).

Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml:64]: Sorry, I haven't added any files. done

Comment (llewelld) [reply apps/browser/qml/pages/components/TabView.qml:64]: Thank you! You should be safe to remove the `_changeMenuApparentHeight()` logic now, and replace the header topMargin to be a fixed height. I'll add a comment about it above.

File: apps/browser/qml/pages/components/TabView.qml
Comment (llewelld) [inline apps/browser/qml/pages/components/TabView.qml:64]: This should now be safe:
```suggestion
                    topMargin: _pulleyMenuHeight
```
Which would allow you to remove the `_changeMenuApparentHeight()` logic I think.

Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: I agree, removed

**Review by llewelld — APPROVED** [review]
Thanks for your work on this, it looks good to me, and is working really great in my opinion.

Please rebase before we merge it.

**Comment by xlise** [comment]
> Thanks for your work on this, it looks good to me, and is working really great in my opinion.
> 
> Please rebase before we merge it.

Yes, it works, but I still haven't corrected one comment from Andrew, but it doesn't work there as Andrew suggested

**Comment by llewelld** [comment]
Thanks, and understood. I won't merge anything until you've completed your discussion with Andrew and you say it's ready (it still needs approval from someone at OMP too before we can merge it).

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: Thank you, this option is really better. done

File: apps/browser/qml/pages/components/TabView.qml
Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: done

**Review by adenexter — COMMENTED** [review]
> Yes, it works, but I still haven't corrected one comment from Andrew, but it doesn't work there as Andrew suggested. Rebased

Which comment was this? The only thing I can see that might be unresolved is that you could use a delegate and model for the Silica TabView the same as with the header so as to not have two mostly the same TabItem components.

File: apps/browser/qml/pages/components/TabView.qml
Comment (adenexter) [reply apps/browser/qml/pages/components/TabView.qml:71]: Correction, Silica TabView has a tabBarHeight property so you can use that instead of defining your own headerHeight property.

Comment (xlise) [reply apps/browser/qml/pages/components/TabView.qml]: Thanks!  done

**Comment by xlise** [comment]
> > Yes, it works, but I still haven't corrected one comment from Andrew, but it doesn't work there as Andrew suggested. Rebased
> 
> Which comment was this? The only thing I can see that might be unresolved is that you could use a delegate and model for the Silica TabView the same as with the header so as to not have two mostly the same TabItem components.

Yes, I mean this remark, I corrected the TabBar. delegate for components today I will try to make

**Comment by adenexter** [comment]
> Yes, I mean this remark, I corrected the TabBar. delegate for components today I will try to make

Okay. Something to watch for is TabGridView has both model and privateMode properties, so if you try and reference the model data from within the TabGridView declaration both model.privateMode and privateMode will resolve to its properties and not the model data.

The best way to address that is probably bind model.privateMode to a property in the delegate top level TabItem and reference that.

**Comment by xlise** [comment]
> > Yes, I mean this remark, I corrected the TabBar. delegate for components today I will try to make
> 
> Okay. Something to watch for is TabGridView has both model and privateMode properties, so if you try and reference the model data from within the TabGridView declaration both model.privateMode and privateMode will resolve to its properties and not the model data.
> 
> The best way to address that is probably bind model.privateMode to a property in the delegate top level TabItem and reference that.

Great thanks a lot for your hint. The code is much better. done

**Review by adenexter — APPROVED** [review]

**Review by atatarov — APPROVED** [review]

**Comment by aroditelev** [comment]
LGTM

**Pull request merged**
